### ISSUE TITLE
Analyzer does not reject raw types in the factory constructor signature

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#37037]
Dart SDK Version: 2.3.1-dev.0.0
OS: Windows 10 64 bit

The following source code example throws compile error with dart, but analyzer does not show any error or warning here:
``` dart
class A<X extends A<X>> {
  A() {}
  factory A.foo() = C;
}

class C<X extends A<X>> extends A<X> {
  C() {}
}

main() {}
```
Seems like dart and analyzer should behave in the similar way.

Sample output is:
> $> dartanalyzer test.dart
> Analyzing test.dart...
> No issues found!
> 
> $> dart test.dart
> test.dart:3:21: Error: The type 'dynamic' doesn't extend 'A<dynamic>'.
>  - 'A' is from 'test.dart'.
> Try using a different type as argument.
>   factory A.foo() = C;

**Comment by eernstg** [comment]
Let's restate the example with different type variable names, just to avoid some possible confusion:

```dart
class A<X extends A<X>> {
  A() {}
  factory A.foo() = C;
}

class C<Y extends A<Y>> extends A<Y> {
  C() {}
}
```

The raw `C` in the body of `A.foo` is an [error](https://github.com/dart-lang/language/blob/0e6111c01e7f35ad358c01570e5c865bd5269ce3/specification/dartLangSpec.tex#L3372):

```
It is a compile-time error if $[T_1/X_1, \ldots, T_m/X_m]F'$
is not a subtype of the function type of $k$.
```

In the context `$k$` is the redirecting constructor `A.foo`, whose function type is `A<X> Function()` (a constructor isn't a function with that type, but it's a convenient terminology which is defined in Sec. 10.6); `$k'$` is the redirectee `C`, and we take that to be raw such that it is transformed into `C<A<dynamic>>` by instantiation to bound. `F'` is the function type of `$k'$`, that is, `C<Y> Function()`. So `[A<dynamic>/Y]F'` is `C<A<dynamic>> Function()`. So we get the error because the required subtype relationship does not exist (in particular, `C<A<dynamic>> <\: A<X>`).

However, it is not completely obvious whether we'd apply _type inference_ to the redirection `C`, in which case it could be taken to mean `C<X>`, and in that case we'd need to check for `[X/Y]F'` = `C<X> Function()` being a subtype of `A<X> Function()`, which succeeds.

Leaf's proposed [specification of type inference](https://github.com/dart-lang/language/blob/3b46550bd069bb62c2f30300039711f886741edb/resources/type-system/inference.md) does not mention the locations where type inference is applied.

If we perform the i2b manually,

```dart
class A<X extends A<X>> {
  A() {}
  factory A.foo() = C<A<dynamic>>;
}

class C<X extends A<X>> extends A<X> {
  C() {}
}
```

then the analyzer (from c4e8cf469a0ff7c19fab0578b496b3593b17a1d5) complains as follows:

```
ERROR|STATIC_WARNING|REDIRECT_TO_INVALID_RETURN_TYPE|/usr/local/google/home/eernst/lang/dart/scratch/201905/n013.dart|3|21|13|The return type 'C<A<dynamic>>' of the redirected constructor isn't assignable to 'A<X>'.
ERROR|STATIC_TYPE_WARNING|TYPE_ARGUMENT_NOT_MATCHING_BOUNDS|/usr/local/google/home/eernst/lang/dart/scratch/201905/n013.dart|3|23|10|'A' doesn't extend 'A<A>'.
```

which is correct: The returned object does not have the promised type `A<X>`, and it is an error to have a super-bounded type as the body of a redirecting factory.

So the analyzer might actually be doing the right thing, based on a decision to apply type inference to `C`.

**Comment by eernstg** [comment]
Confirmed, the analyzer _does_ perform type inference on the redirection of the factory:

```dart
class A<X extends num> {
  const factory A() = B;
}

class B<X extends num> implements A<X> {
  const B();
}

const a = A<int>();
const b1 = a is B<num>;
const b2 = a is B<int>;
const b3 = a is B<double>;

class C {
  final int i;
  const C(): i = 1 ~/ (b2 ? 0 : 1);
}

main() {
  const C(); // Trigger the compile-time computation of i.
}
```

Using `b2 ? 0 : 1` as shown we get a compile-time error:

```
ERROR|COMPILE_TIME_ERROR|CONST_EVAL_THROWS_EXCEPTION|/usr/local/google/home/eernst/lang/dart/scratch/201905/n015.dart|21|3|9|Evaluation of this constant expression throws an exception.
```

Replacing that expression by `b1 ? 0 : 1` yields the error (as expected), but `b3 ? 0 : 1` does not. Given that `a` is a `B<int>` and not a `B<double>` (so it's not a `B<Null>`), we must have that the dynamic type of `a` is exactly `B<int>`, and this implies that the analyzer uses type inference on the factory redirection.

**Comment by eernstg** [comment]
I just created https://github.com/dart-lang/language/issues/366, proposing that we use type inference on redirecting factory constructor bodies which are raw types (or `R.id` where `R` is a raw type).

**Comment by iarkh** [comment]
Seems like analyzer behaves correctly and this Issue should be closed, see Issue #34714 for more details.
I've just filed an Issue #41895 against runtime.

**Comment by eernstg** [comment]
Agreed, the analyzer should and does use inference, so we can close this issue. Note that #34714 was closed because this inference step was implemented in the common front end in https://dart-review.googlesource.com/c/sdk/+/147181.
